geoapps/octree_creation/driver.py

```diff
@@ -77,46 +80,174 @@ def octree_from_params(params: OctreeParams):
 
+        if isinstance(levels, list):
+            levels = np.array(levels)
+
+        distance = 0
         for ii, n_cells in enumerate(levels):
             distance += n_cells * treemesh.h[0][0] * 2**ii
 
             treemesh.refine_ball(
-                locs, distance, treemesh.max_level - ii, finalize=finalize
+                locs, distance, treemesh.max_level - ii, finalize=False
             )
 
+        if finalize:
+            treemesh.finalize()
+
+        return treemesh
+
+    @staticmethod
+    def refine_tree_from_surface(
+        treemesh: TreeMesh,
+        surface: ObjectBase,
+        levels: list[int] | np.ndarray,
+        max_distance: float = np.inf,
+        finalize=False,
+    ):
+        """
+        Refine a tree mesh along the simplicies of a surface.
+
+        :param treemesh: Tree mesh to refine.
+        :param surface: Surface object to use for refinement.
+        :param levels: Number of cells requested at each refinement level.
+            Defined in reversed order from highest octree to lowest.
+        :param max_distance: Maximum distance from the surface to refine.
+        :param finalize: Finalize the tree mesh after refinement.
+
+        :return: Refined tree mesh.
+        """
+        if isinstance(levels, list):
+            levels = np.array(levels)
+
+        xyz = get_locations(surface.workspace, surface)
+        tri2D = Delaunay(xyz[:, :2])
+        tree = cKDTree(xyz[:, :2])
+
+        if isinstance(surface, Surface):
+            tri2D.simplices = surface.cells
+
+        F = interpolate.LinearNDInterpolator(tri2D, xyz[:, -1])
+        min_locs = np.min(xyz, axis=0)
+        max_locs = np.max(xyz, axis=0)
+        levels = np.array(levels)
+
+        depth = 0
+        # Cycle through the Tree levels backward
+        for ind, n_cells in enumerate(levels):
+            if n_cells == 0:
+                continue
+
+            dx = treemesh.h[0][0] * 2**ind
+            dy = treemesh.h[1][0] * 2**ind
+            dz = treemesh.h[2][0] * 2**ind
+
+            # Create a grid at the octree level in xy
+            CCx, CCy = np.meshgrid(
+                np.arange(min_locs[0], max_locs[0], dx),
+                np.arange(min_locs[1], max_locs[1], dy),
+            )
+            xy = np.c_[CCx.reshape(-1), CCy.reshape(-1)]
+
+            # Only keep points within triangulation
+            inside = tri2D.find_simplex(xy) != -1
+            r, _ = tree.query(xy)
+            keeper = np.logical_and(r < max_distance, inside)
+            nnz = keeper.sum()
+            elevation = F(xy[keeper])
+
+            # Apply vertical padding for current octree level
+            for _ in range(int(n_cells)):
+                depth += dz
+                treemesh.insert_cells(
+                    np.c_[xy[keeper], elevation - depth],
+                    np.ones(nnz) * treemesh.max_level - ind,
+                    finalize=False,
```

Comment:
again, why not just use the inpute argument since it's the last operation?

Reply:
Same